src/threads-version/big_int64.cpp

```
@@ -0,0 +1,767 @@
+                      static_cast<uint64_t>(str[index - 7]));
+    }
+    if (msbParts != 0) {
+      uint64_t msb = 0;
+      for (int i = 0; i < msbParts; i++) {
+        msb |= static_cast<uint64_t>(str[i]) << (i * BITS_IN_BYTE);
+      }
+      limbs.push_back(msb);
+    }
+  } else {
+    // big endian-> {0,1,2,3,4,5,6,7} becones 0x 00 01 02 03 04 05 06 07
+    for (int i = 0; i < fullLimbs; i++) {
+      int index = strLen - 1 - i * BYTES_IN_LIMB;
+      limbs.push_back(static_cast<uint64_t>(str[index]) |
+                      (static_cast<uint64_t>(str[index - 1]) << 8) |
+                      (static_cast<uint64_t>(str[index - 2]) << 16) |
+                      (static_cast<uint64_t>(str[index - 3]) << 24) |
+                      (static_cast<uint64_t>(str[index - 4]) << 32) |
+                      (static_cast<uint64_t>(str[index - 5]) << 40) |
+                      (static_cast<uint64_t>(str[index - 6]) << 48) |
+                      (static_cast<uint64_t>(str[index - 7]) << 56));
+    }
+    if (msbParts != 0) {
+      uint64_t msb = 0;
+      for (int i = 0; i < msbParts; i++) {
+        msb |= static_cast<uint64_t>(str[msbParts - 1 - i])
+               << (i * BITS_IN_BYTE);
+      }
+      limbs.push_back(msb);
+    }
+  }
+  removeLeadingZeros();
+}
+
+BigInt64::BigInt64(std::vector<uint64_t> v, bool isNegative)
+    : limbs(v), isNegative(isNegative) {}
+
+BigInt64::BigInt64(CreateModes mode, int bits) : isNegative(false) {
+  if (bits >= BITS_IN_LIMB)
+    generateNLimbsRandom(bits / BITS_IN_LIMB);
+  bits = bits % BITS_IN_LIMB; // 0-63
+  if (bits != 0) {
+    uint64_t one = static_cast<uint64_t>(1);
+    limbs.push_back(randomLimb(0, (one << bits) - 1) | (one << (bits - 1)));
+  } else
+    limbs.back() |= 0x8000000000000000ULL;
+}
+
+BigInt64::BigInt64(CreateModes mode, uint64_t min, const BigInt64 &max)
+    : isNegative(false) {
+  if (max.isSmallerThanUnsigned(min))
+    throw invalid_argument("min must be <= max");
```

Should return an error code?

---

wrapped with try catch